pypowerwall/__init__.py

```diff
@@ -581,45 +581,47 @@ def alerts(self, jsonformat=False, alertsonly=True) -> Union[list, str]:
           jsonformat = If True, return JSON format otherwise return Python Dictionary
           alertonly  = If True, return only alerts without device name
         """
-        alerts = []
+        alerts = set()
         devices: dict = self.vitals() or {}
         if devices:
             for device in devices:
-                if 'alerts' in devices[device]:
-                    for i in devices[device]['alerts']:
-                        if alertsonly:
-                            alerts.append(i)
-                        else:
-                            item = {device: i}
-                            alerts.append(item)
+                if not 'alerts' in devices[device]:
+                    continue
+                for i in devices[device]['alerts']:
+                    if alertsonly:
+                        alerts.add(i)
+                        continue
+                    alerts.add({device: i})
         elif not devices and alertsonly is True:
             # Vitals API is not present in firmware versions > 23.44 for local mode.
             # This is a workaround to get alerts from the /api/solar_powerwall endpoint
             data: dict = self.poll('/api/solar_powerwall') or {}
             pvac_alerts = data.get('pvac_alerts') or {}
             for alert, value in pvac_alerts.items():
                 if value is True:
-                    alerts.append(alert)
+                    alerts.add(alert)
             pvs_alerts = data.get('pvs_alerts') or {}
             for alert, value in pvs_alerts.items():
                 if value is True:
-                    alerts.append(alert)
+                    alerts.add(alert)
 
         # Augment with inferred alerts from the grid_status
         grid_status = self.poll('/api/system_status/grid_status')
         if grid_status:
             alert = grid_status.get('grid_status')
-            if alert == 'SystemGridConnected' and 'SystemConnectedToGrid' not in alerts:
-                alerts.append('SystemConnectedToGrid')
-            elif alert:
-                alerts.append(alert)
             if grid_status.get('grid_services_active'):
-                alerts.append('GridServicesActive')
+                alerts.add('GridServicesActive')
+            elif alert:
+                alerts.add(alert)
+
+        def normalize_alerts(s: str) -> str:
+            return s.replace("SystemGridConnected", "SystemConnectedToGrid")
+
+        normalized_alerts = {normalize_alerts(s) for s in alerts}
```

Comment:
Instead of creating a function, why not do this?

```python
        normalized_alerts = {s.replace("SystemGridConnected", "SystemConnectedToGrid") for s in alerts}
```

Reply:
Because I wanted to leave room for more substitutions in the future. However, it is premature optimization and does increase the complexity of the function needlessly, so I'll keep it simple.